enhancements/network/static-ip-addresses-vsphere.md

```diff
@@ -0,0 +1,564 @@
+                config:
+                  server:
+                    - 192.168.1.215
+              routes:
+                config:
+                  - next-hop-address: 192.168.100.1
+      metadata:
+        creationTimestamp: null
+      numCPUs: 2      
+      kind: VSphereMachineProviderSpec
+      workspace:
+        datacenter: testdc
+        datastore: datastore-1
+        folder: /testdc/vm/cluster-folder
+        resourcePool: /testdc/host/cluster1/Resources
+        server: test.vcenter.net
+      template: vm-template-rhcos
+      apiVersion: machine.openshift.io/v1beta1
+~~~
+4. For bootstrap and control plane nodes, modify vSphere terraform to convert nmstate to a VM guestinfo parameter
+for each VM to be created.
+
+As the assets are generated for the control plane and compute nodes, the slice of `host`s for each 
+node role will be used to populate `nmstate` information.  The number of `host`s must match the number of
+replicas defined in the associated machine pool.
+
+Additionally, each defined host may optionally define a failure domain.  This indicates that the associated `networkConfig` will be applied to a machine created in the indicated failure domain.
+
+
+##### Machine API
+- Modify vSphere machine controller to convert nmstate to VM guestinfo parameter
+- Introduce a new lifecycle hook called `preCreate`.
+- Modify [types_vsphereprovider.go](https://github.com/openshift/api/blob/master/machine/v1beta1/types_vsphereprovider.go) to support nmstate configuration. 
+
+
+###### IP Configuration of Machines
+The machine API `VSphereMachineProviderSpec.Network` will be modified to include a new type called `NetworkConfig`.  
+
+~~~go
+// NetworkDeviceSpec defines the network configuration for a virtual machine's
+// network device.
+type NetworkDeviceSpec struct {
+	// NetworkName is the name of the vSphere network to which the device
+	// will be connected.
+	NetworkName string `json:"networkName"`
+
+	// Config specifies an IP configuration to be applied upon cloning.
+	// +optional
+	Config *NetworkConfig `json:"config,omitempty"`
+}
+~~~
+
+See [installer](#installer) for a definition of this type as well as other types that make up the definition of `NetworkConfig`.  `NetworkConfig` and it's dependent types will be defined in [types_vsphereprovider.go](https://github.com/openshift/api/blob/master/machine/v1beta1/types_vsphereprovider.go).
+
+An additional lifecycle hook will be added to the The `machine.machine.openshift.io` and `machinesets.machine.openshift.io` CRDs to enable a controller to augment a machine resource before it is rendered in the backing infrastructure.
+
+~~~go
+// PreCreate hooks prevent the machine from being created in the backing infrastructure.
+// +listType=map
+// +listMapKey=name
+// +optional
+PreCreate []LifecycleHook `json:"preCreate,omitempty"`
+~~~
+
+Creation of the resource will be blocked until the `preCreate` hook is removed from the `machine.machine.openshift.io` instance.
```

Comment:
When there's a precreate hook, does this prevent the Machine from going into the `provisioning` phase? I'd expect so. We will also want to add a condition (mirroring what's in the other deletion hooks), can we add a note here about that as well. Do you plan to update the deletion hooks enhancement?

Reply:
In my prototype, the machine transitions to `provisioning` and waits for the `preCreate` hooks to be removed.  However, that was just to proof out the idea.  I will add details dictating that the machine shall not enter the `provisioning` phase until the `preCreate` hooks are removed.  

I was planning on carrying details on the hook in this enhancement, but it probably makes more sense to update https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/machine-deletion-hooks.md .  I'll start a PR to update that enhancement.

Reply:
A pull request has been started for the `preCreate` lifecycle hook. https://github.com/openshift/enhancements/pull/1273

Reply:
A pull request has been started for the nmstate related API changes https://github.com/openshift/api/pull/1338 .  I was torn on where the nmstate types should live so any feedback on that is much appreciated.

Reply:
>  the machine transitions to provisioning and waits for the preCreate hooks to be removed.

Reply:
I do think that would make it clearer.

Reply:
I can go ahead and make that update if you agree it would be clearer and is a direction we want to go.